### PR TITLE
Fix leaked Vault LifetimeRenewers

### PR DESCRIPTION
**Pull request by clly**
When the Vault CA Provider is reconfigured we do not stop the
LifetimeRenewers which can cause them to leak until the Consul processes
recycles. On Configure execute stopWatcher if it exists and is not nil
before starting a new renewal

Also add jitter before restarting the LifetimeWatcher to prevent overwhelming Vault
when it's configured with rate limiting.

If we fail to login to Vault or our token is no longer valid we can
overwhelm a Vault instance with many requests very quickly by restarting
the LifetimeWatcher. Before restarting the LifetimeWatcher provide a
backoff time of 1 second or less.

### CODE REVIEW

File: agent/connect/ca/provider_vault.go
Comment (banks) [inline agent/connect/ca/provider_vault.go:186]: What was the change on Vault side? I see backoff in their code from over a year ago.

If we upgrade to 1.1.1, do we still need to sleep here? If so we could maybe use an exponential backoff too - it's still a bit gross to be making on average 2 failing auth calls per second forever if we have a bad token. I'd probably use our `lib.Retry` for it in `/lib`.

Finally, is there something stopping us from upgrading to api@1.1.1 now?

**Review by banks — COMMENTED** [review]
Thanks @clly this looks pretty good and may be OK to merge as-is. I had a few questions inline though.

Mostly trying to work out if the hot loop was caused by Vault's API code (so the real fix is to upgrade) or if Vault's API code is already doing backoff internally and the hot loop we saw was actually caused by this code because the lifetime watcher is failing fast but we are immediately restarting it?

If it's the latter then I think we should maybe be more aggressive with backoff here as it's not just an edgecase but a real error that can happen easily!

File: agent/connect/ca/provider_vault.go
Comment (clly) [reply agent/connect/ca/provider_vault.go:186]: The change of the Vault side _was_ the addition of the exponential backoff from over a year ago. If we upgrade to 1.1.1 we do not need to sleep here because the watcher should get an error, retry with the exponential backoff, then exit. The Vault code is currently returning attempting to renew, returning quickly and we are immediately restarting it causing another renew attempt.

Upgrading to api@1.1.1 upgrades some other dependencies including protobuf, the aws-sdk, and genproto. I'm not super familiar with what those would have affect or what our normal process is for upgrading our dependencies so I put the sleep here instead. The exponential retry in the Vault code would cap out at 1 second so we'd still have 1 failing call per second forever (or until consul is reconfigured).

I can switch to lib.Retry fairly easily or upgrade to the newer vault/api version.

Comment (banks) [reply agent/connect/ca/provider_vault.go:186]: Hmm yeah the protobuf thing probably means we can't until we are done moving off gogo but I'll defer to the Core team to work out if that's true!

I think this is OK for a temporary solution but maybe we could:
 - make it `lib.RandomJitter(4 *time.Second) + time.Second` so it's 1-5 seconds rather than 0-1. (or lib.Retry which is not too hard?) either seems fine
 - Add `TODO` comments here and maybe link this PR in the issue we are tracking protobuf upgrades so we don't forget to come back and upgrade?

I think it's worth trying to merge and ship this fix as soon as possible as it could affect production Vault stability.

@hashicorp/consul-core, what do you think?

Comment (clly) [reply agent/connect/ca/provider_vault.go]: I've switched from RandomJitter to lib.Retry. I've also created a separate issue to upgrade to a newer version of the vault/api and linked it here and to https://github.com/hashicorp/consul/issues/10471

File: agent/connect/ca/provider_vault.go
Comment (banks) [inline agent/connect/ca/provider_vault.go:265]: Does this get called after every successful request to vault from the watcher? For example does it fire after the initial token fetch/validation? If not, we might need to also reset above when we re-start the watcher?

File: agent/connect/ca/provider_vault_test.go
Comment (banks) [inline agent/connect/ca/provider_vault_test.go:238]: Is this callback will called from another goroutine? I think it is, which probably means this would fail the race detector (even though bool assign is atomic on some CPUs anyway).

Could replace it with:
```suggestion
	var gotStopped = uint32(0)
	provider.stopWatcher = func() {
		atomic.StoreUint32(&gotStopped, 1)
	}
```

then later `require.Equal(t, uint32(1),  atomic.LoadUint32(&gotStopped))`

**Review by banks — APPROVED** [review]
A couple of minor things but LGTM. Would be great if someone from Core could review too!

File: agent/connect/ca/provider_vault.go
Comment (clly) [reply agent/connect/ca/provider_vault.go:265]: This gets called on every successful renewal which should happen immediately after start and then the LifetimeWatcher sleeps.

File: agent/connect/ca/provider_vault_test.go
Comment (clly) [reply agent/connect/ca/provider_vault_test.go:238]: It shouldn't be called as part of a goroutine (at least in this test) but is a call back from `provider.Configure.` Using atomic should make it more robust if that changes in the future.

**Review by kyhavlov — APPROVED** [review]
LGTM 👍

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/615051.

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit 922619dfc33572c4f6076fae7e7dce0ec5fe450c onto `release/1.11.x` succeeded!

**Comment by hc-github-team-consul-core** [comment]
:cherries::x: Cherry pick of commit 922619dfc33572c4f6076fae7e7dce0ec5fe450c onto `release/1.10.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/615051)

**Pull request merged**
